koin-core/src/main/kotlin/org/koin/KoinContext.kt

```diff
@@ -71,27 +79,41 @@ class KoinContext(
 
         val clazzName = clazz.java.canonicalName
 
-        var resultInstance: T? = null
-
         val beanDefinition: BeanDefinition<*> =
             getVisibleBeanDefinition(clazzName, definitionResolver, resolutionStack.last())
 
-        val logIndent = resolutionStack.indent()
-        resolutionStack.resolve(beanDefinition) {
+        val instance = resolveBeanDefinition<T>(clazzName, beanDefinition, parameters, resolutionStack.indent())
 
-            // Resolution log
-            logger.log("${logIndent}Resolve class[$clazzName] with $beanDefinition")
+        return instance ?: error("Could not create instance for $clazz")
+    }
 
-            val (instance, created) = instanceFactory.retrieveInstance<T>(beanDefinition, ParameterHolder(parameters))
+    /**
+     * Resolve all dependencies for their bean definitions
+     * @param clazz - Class
+     * @param parameters - Parameters
+     * @param definitionResolver - function to find bean definitions
+     */
+    fun <T> resolveAllInstances(
+        clazz: KClass<*>,
+        parameters: Parameters,
+        definitionResolver: () -> List<BeanDefinition<*>>
+    ): Set<T> = synchronized(this) {
 
-            // Log creation
-            if (created) {
-                logger.log("$logIndent(*) Created")
-            }
+        val clazzName = clazz.java.canonicalName
 
-            resultInstance = instance
+        val instances: MutableSet<T> = HashSet()
+
+        val beanDefinitions = getAllVisibleBeanDefinitions(clazzName, definitionResolver, resolutionStack.last())
+
+        val logIndent = resolutionStack.indent()
+        for (beanDefinition in beanDefinitions) {
```

Comment:
You should be able to use `mapNotNull` here instead (see https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/map-not-null.html )
Something like
```
return beanDefinitions.mapNotNull { resolveBeanDefinition<T>(clazzName, it, parameters, logIndent) }.toSet()
```

Or `mapNotNullTo` if you want to collect in a set directly.

Reply:
Maybe null result from `resolveBeanDefinition` shouldn't be ignore be raise an exception, as done in the "single" version of the method. Ignoring them could lead to nasty behaviors at runtime. Will need to ask @arnaudgiuliani to be sure.